File: Google/gcj/17/flipper/pancake_flipper.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

int minimum(int a, int b) {
    return (a <= b) ? a : b;
}

int solve(string s, int k) {
    int cnt = 0;
    for(int i=0; i<=s.length()-k; i++) {
        if(s[i] == '-') {
            for(int j=i; j<i+k; j++) {
                if(s[j] == '+') s[j] = '-';
                else s[j] = '+';
            }
            cnt += 1;
        }
    }
    for(int i=0; i<s.length(); i++) {
        if(s[i] == '-') return -1;
    }
    return cnt;
}

int main() {
    int testcases;
    cin >> testcases;
    for(int test =1 ; test <= testcases; test++) {
        string n;
        int k;
        cin >> n >> k;
        int result = solve(n, k);
        if(result == -1) 
            printf("Case #%d: IMPOSSIBLE\n", test);
        else
            printf("Case #%d: %d\n", test, result);
    }
    return 0;
}

File: Google/gcj/17/flipper/pancake_flipper.py
import pdb

class Queue:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return self.items == []

    def enqueue(self, item):
        self.items.insert(0,item)

    def dequeue(self):
        return self.items.pop()

    def size(self):
        return len(self.items)

    def get(self):
        return self.items

def find_flips(pancakes):
    indices = []
    for i in range(0, len(pancakes)):
        if pancakes[i] == '-': indices.append(i)
    return indices

def no_more_flips(pancakes):
    for i in range(0, len(pancakes)):
        if pancakes[i] == '-': return False
    return True

def is_again_same_thing(pancakes, res):
    res1 = ''.join(pancakes)
    if res1 == res: return True
    return False

def flip_pancakes(pancakes, pos):
    for value in pos:
        if value < 0 or value >= len(pancakes): return pancakes
    new_pancakes = []
    for i in range(0, len(pancakes)):
        if i not in pos: new_pancakes.append(pancakes[i])
        else:
            if pancakes[i] == '+': new_pancakes.append('-')
            else: new_pancakes.append('+')
    return new_pancakes

def get_positions(index, window):
	positions = []
	sindex = index - (window - 1)
	for win in range(0, window):
		positions.append([i for i in range(sindex, sindex+window)])
		sindex = sindex + 1
	return positions

def solve(pancakes, result, window):
	que = Queue()
	que.enqueue((pancakes, 0, []))
	visited = {}
        while not que.isEmpty():
		temp, steps, pos = que.dequeue()
                if visited.get(''.join(temp), False): continue
		if no_more_flips(temp): return steps
		flips = find_flips(temp)
		for flip in flips:
			positions = get_positions(flip, window)
			for pos in positions:
				changed_pancakes = flip_pancakes(temp, pos)
				if is_again_same_thing(changed_pancakes, result): continue
				que.enqueue((changed_pancakes, steps+1, pos))
                visited[''.join(temp)] = True
        return -1

if __name__ == "__main__":
    testcases = int(raw_input())
    for t in range(1, testcases+1):
        pancakes, window = raw_input().split()
        pancakes, window = list(pancakes), int(window)
        result = solve(pancakes, ''.join(pancakes), window)
        if result != -1:
        	print "Case #%d: %d" %(t, result)
        else:
        	print "Case #%d: %s" %(t, "IMPOSSIBLE")
